### PR TITLE
Additional optimisations to thumbnail generation and quality

### PR DESCRIPTION
**Pull request by theefer**
Will comment inline.

Merges into the existing thumbnail PR.

### CODE REVIEW

File: common-lib/src/main/scala/com/gu/mediaservice/lib/imaging/ImageOperations.scala [hunk 1]
```diff
@@ -102,19 +102,18 @@ object ImageOperations {
...
   val thumbUnsharpSigma = 0.5d
   val thumbUnsharpAmount = 0.8d
-  val thumbUnsharpThreshold = 0.02d
   def createThumbnail(sourceFile: File, width: Int, qual: Double = 100d, tempDir: File, iccColourSpace: Option[String], colourModel: Option[String]): Future[File] = {
     for {
       outputFile <- createTempFile(s"thumb-", ".jpg", tempDir)
       cropSource  = addImage(sourceFile)
-      qualified   = quality(cropSource)(qual)
-      corrected   = correctColour(qualified)(iccColourSpace, colourModel)
+      thumbnailed = thumbnail(cropSource)(width)
```
Comment (theefer) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/imaging/ImageOperations.scala]: `thumbnail` is significantly faster than `resize`, although not quite as good in terms of quality, but good enough for thumbnails.

File: common-lib/src/main/scala/com/gu/mediaservice/lib/imaging/ImageOperations.scala [hunk 2]
```diff
@@ -102,19 +102,18 @@ object ImageOperations {
...
+      corrected   = correctColour(thumbnailed)(iccColourSpace, colourModel)
       converted   = applyOutputProfile(corrected, optimised = true)
       stripped    = stripMeta(converted)
       profiled    = applyOutputProfile(stripped, optimised = true)
-      resized     = resize(profiled)(width)
-      unsharpened = unsharp(resized)(thumbUnsharpRadius, thumbUnsharpSigma, thumbUnsharpAmount, thumbUnsharpThreshold)
-      addOutput   = addDestImage(unsharpened)(outputFile)
+      unsharpened = unsharp(profiled)(thumbUnsharpRadius, thumbUnsharpSigma, thumbUnsharpAmount)
```
Comment (theefer) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/imaging/ImageOperations.scala]: Sharpening significantly improves the look of thumbnails.

File: common-lib/src/main/scala/com/gu/mediaservice/lib/imaging/ImageOperations.scala (hunk 1 above)
Comment (theefer) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/imaging/ImageOperations.scala]: Now sizing down *before* doing any of the expensive operations, especially colourspace conversion, to make them cheaper (work on less data).

File: image-loader/app/lib/Config.scala [hunk 3]
```diff
@@ -26,7 +26,7 @@ object Config extends CommonPlayAppProperties with CommonPlayAppConfig {
   val tempDir: File = new File(properties.getOrElse("upload.tmp.dir", "/tmp"))
 
   val thumbWidth: Int = 256
-  val thumbQuality: Double = 75d // out of 100
+  val thumbQuality: Double = 85d // out of 100
```
Comment (theefer) [inline image-loader/app/lib/Config.scala]: Raised quality to reduce artifacts. This is offset by other optimisations, esp stripping metadata and injecting much smaller profile data, which significantly reduces the size even with raised quality.

**Comment by jamesgorrie** [comment]
Out of interest - what savings are we making (filesize and speed)?

File: common-lib/src/main/scala/com/gu/mediaservice/lib/imaging/ImageOperations.scala (hunk 2 above)
Comment (jamesgorrie) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/imaging/ImageOperations.scala]: `unsharped` what a silly name.

**Comment by jamesgorrie** [comment]
Looks good! :+1:

**Comment by theefer** [comment]
Speed-wise in terms of latency of image loader, seems very similar to the current performance:

image
(Before 10am is using the new processing from this branch, after 10am is back to the old/current processing)

Filesize I don't have aggregate numbers, but looks around 20 KB on average for the old/current system vs 12 KB for the new system, but with a much larger variance in the current system (likely because we left the profile data in, which can be large).

File: common-lib/src/main/scala/com/gu/mediaservice/lib/imaging/ImageOperations.scala (hunk 2 above)
Comment (paperboyo) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/imaging/ImageOperations.scala]: > `unsharped` what a silly name.

[Mixes in a blurred original](https://en.wikipedia.org/wiki/Unsharp_masking), that's why...

**Pull request merged**